base/applications/rapps/cabinet.cpp

```diff
@@ -0,0 +1,351 @@
+                             0);
+    }
+    break;
+    case fdintCLOSE_FILE_INFO:
+        // Ignore all props and just close the file
+        iResult = !fnFileClose(pfdin->hf) ? TRUE : FALSE;
+        break;
+    case fdintNEXT_CABINET:
+        if (pfdin->fdie != FDIERROR_NONE)
+        {
+            iResult = -1;
+        }
+        break;
+    case fdintPARTIAL_FILE:
+        iResult = 0;
+        break;
+    case fdintCABINET_INFO:
+        iResult = 0;
+        break;
+    case fdintENUMERATE:
+        iResult = 0;
+        break;
+    default:
+        iResult = -1;
+        break;
+    }
+
+    return iResult;
+}
+
+/* cabinet.dll FDI function pointers */
+
+typedef HFDI(*fnFDICreate)(PFNALLOC, 
+                           PFNFREE, 
+                           PFNOPEN, 
+                           PFNREAD, 
+                           PFNWRITE,
+                           PFNCLOSE, 
+                           PFNSEEK, 
+                           int, 
+                           PERF);
+
+typedef BOOL(*fnFDICopy)(HFDI,
+                         LPSTR,
+                         LPSTR,
+                         INT,
+                         PFNFDINOTIFY,
+                         PFNFDIDECRYPT,
+                         void FAR *pvUser);
+
+typedef BOOL(*fnFDIDestroy)(HFDI);
+
+/* 
+ * Extraction function 
+ * TODO: require only a full path to the cab as an argument
+ */
+BOOL ExtractFilesFromCab(const ATL::CStringW& szCabName, 
+                         const ATL::CStringW& szCabDir, 
+                         const ATL::CStringW& szOutputDir)
+{
+    HINSTANCE hCabinetDll;
+    HFDI ExtractHandler;
+    ERF ExtractErrors;
+    ATL::CStringA szCabNameUTF8, szCabDirUTF8, szOutputDirUTF8;
+    fnFDICreate pfnFDICreate;
+    fnFDICopy pfnFDICopy;
+    fnFDIDestroy pfnFDIDestroy;
+    BOOL bResult;
+
+    // Load cabinet.dll and extract needed functions 
+    hCabinetDll = LoadLibraryW(L"cabinet.dll");
+
+    if (!hCabinetDll)
+    {
+        return FALSE;
+    }
+
+    pfnFDICreate = (fnFDICreate) GetProcAddress(hCabinetDll, "FDICreate");
+    pfnFDICopy = (fnFDICopy) GetProcAddress(hCabinetDll, "FDICopy");
+    pfnFDIDestroy = (fnFDIDestroy) GetProcAddress(hCabinetDll, "FDIDestroy");
+
+    if (!pfnFDICreate || !pfnFDICopy || !pfnFDIDestroy)
+    {
+        FreeLibrary(hCabinetDll);
+        return FALSE;
+    }
+
+    // Create FDI context
+    ExtractHandler = pfnFDICreate(fnMemAlloc,
+                                  fnMemFree,
+                                  fnFileOpen,
+                                  fnFileRead,
+                                  fnFileWrite,
+                                  fnFileClose,
+                                  fnFileSeek,
+                                  cpuUNKNOWN,
+                                  &ExtractErrors);
+
+    if (!ExtractHandler)
+    {
+        FreeLibrary(hCabinetDll);
+        return FALSE;
+    }
+
+    // Create output dir
+    CreateDirectoryW(szOutputDir, NULL);
```

Comment:
It may be worth looking at the return (success) value of this call, and take appropriate action if something failed.